### ISSUE TITLE
[mesa 19.1.6]dEQP-VK.api.external.memory.android_hardware_buffer.* failed due to crash

### ISSUE THREAD

**Issue by renchenglei** [intel/external-mesa#133]
One vulkan external memory case failed on x86 platform due to "crash" on Android 10 with mesa 19.1.6.
```
dEQP-VK.api.external.memory.android_hardware_buffer.image_formats#r16g16b16a16_sfloat
```
**This issue is also reproduced with mesa 19.0.6. Android framework failed to alloc multi buffers. Then it will report errors. On x86 platform, it seems that the crash is caused by out of heap.**
**issue is also reproduced with upstream master branch(e4a52bd65).**
Steps to reproduce this issue(run with cts packages):
```
run cts -m CtsDeqpTestCases -t dEQP-VK.api.external.memory.android_hardware_buffer.image_formats#r16g16b16a16_sfloat --abi x86
```

**Comment by tpalli** [comment]
This passes for me with Mesa master on KBL NUC, maybe APL specific issue?

**Comment by renchenglei** [comment]
@tpalli, yes, this is APL specific issue. It may be caused by performance issue.

**Comment by renchenglei** [comment]
@tpalli, sorry, maybe I am wrong. This is similar to https://github.com/intel/external-mesa/issues/134, please upgrade minigbm, and have a double try. :)

**Comment by tpalli** [comment]
OK yes, I've reproduced this now on KBL. This seems to trigger 'ubsan: add-overflow', backtrace does not have mesa libraries in it though, this could be an issue within dEQP.

**Comment by tpalli** [comment]
This looks like Android framework issue. For me this test passes if I comment out dumpToSystemLog() call from here:

https://android.googlesource.com/platform/frameworks/native/+/jb-mr0-release/libs/ui/GraphicBuffer.cpp#117

**Comment by renchenglei** [comment]
I remember we have some analysis ago. Android Framework didn't support multi-layer allocate. And it will always failed when we do following:
https://android.googlesource.com/platform/external/deqp/+/refs/heads/deqp-dev/external/vulkancts/modules/vulkan/api/vktApiExternalMemoryTests.cpp#4047
And from AOSP side, it will print the buffer info, which may overflow:
    	1. AHardwareBuffer_allocate(AHardwareBuffer alloc start, call GraphicBufferAllocator::allocate): https://cactus-source.sh.intel.com/source/xref/OneAndroid-MASTER/frameworks/native/libs/nativewindow/AHardwareBuffer.cpp#43
        2. GraphicBufferAllocator::allocate(call allocate to allocate buffer) https://cactus-source.sh.intel.com/source/xref/OneAndroid-MASTER/frameworks/native/libs/ui/GraphicBufferAllocator.cpp#105
    	3. Gralloc2Allocator::allocate(call create descriptor before it call alloc in minigbm) https://cactus-source.sh.intel.com/source/xref/OneAndroid-MASTER/frameworks/native/libs/ui/Gralloc2.cpp#382
    	4. mMapper.createDescriptor(Before it call into minigbm createDescriptor, it seems there is some check from AOSP side) https://cactus-source.sh.intel.com/source/xref/OneAndroid-MASTER/frameworks/native/libs/ui/Gralloc2.cpp#393
    	5. Gralloc2Mapper::createDescriptor(Fail to create descriptor): https://cactus-source.sh.intel.com/source/xref/OneAndroid-MASTER/frameworks/native/libs/ui/Gralloc2.cpp#116
    	6. mMapper->createDescriptor(Call createDescriptor, if layer count is not 1, return error): https://cactus-source.sh.intel.com/source/xref/OneAndroid-MASTER/frameworks/native/libs/ui/Gralloc2.cpp#149
    	7. createDescriptor(return error to GraphicBufferAllocator::allocate): https://cactus-source.sh.intel.com/source/xref/OneAndroid-MASTER/hardware/interfaces/graphics/mapper/2.0/utils/passthrough/include/mapper-passthrough/2.0/Gralloc1Hal.h#77
    	8. return NO_MEMORY(Start to  GraphicBuffer::dumpAllocationsToSystemLog):https://cactus-source.sh.intel.com/source/xref/OneAndroid-MASTER/frameworks/native/libs/ui/GraphicBufferAllocator.cpp#147
    	9. GraphicBuffer::dumpAllocationsToSystemLog(keep calling "dump"): https://cactus-source.sh.intel.com/source/xref/OneAndroid-MASTER/frameworks/native/libs/ui/GraphicBufferAllocator.cpp#98
        10. GraphicBufferAllocator::dump(Crashes with list.valueAt(i)): https://cactus-source.sh.intel.com/source/xref/OneAndroid-MASTER/frameworks/native/libs/ui/GraphicBufferAllocator.cpp#72

**Comment by tpalli** [comment]
I think I have a fix for this one now :)

**Comment by renchenglei** [comment]
@tpalli, great! Wait for the fix!

**Comment by tpalli** [comment]
Here's a hopeful fix, if it works out well I will upstream this for Android.

[0001-libui-use-uint64_t-to-avoid-overflow-issues-with-big.log](https://github.com/intel/external-mesa/files/3717397/0001-libui-use-uint64_t-to-avoid-overflow-issues-with-big.log)

I've tested this only on KBL but there it makes the test pass.

**Comment by renchenglei** [comment]
Thanks, I will have a test on APL(NUC or GP). Will be back asap. :)

**Comment by renchenglei** [comment]
@tpalli, tried with your above changes, and case also could pass on APL NUC. :)

**Comment by tpalli** [comment]
Updated patch ..

[0001-libui-use-uint64_t-in-dump-and-getTotalSize-to-avoid.log](https://github.com/intel/external-mesa/files/3749626/0001-libui-use-uint64_t-in-dump-and-getTotalSize-to-avoid.log)
